test/sample_consensus/test_sample_consensus_line_models.cpp

```diff
@@ -149,6 +149,73 @@ TEST (SampleConsensusModelLine, OnGroundPlane)
   EXPECT_NEAR (0, coeff[5], 1e-4);
 }
 
+TEST (SampleConsensusModelLine, SampleValidationPointsEqual)
+{
+  PointCloud<PointXYZ> cloud;
+  cloud.resize (2);
+
+  cloud[0].getVector3fMap () <<  0.1f,  0.0f,  0.0f;
+  cloud[1].getVector3fMap () <<  0.1f,  0.0f,  0.0f;
+
+  // Create a shared line model pointer directly
+  SampleConsensusModelLinePtr model (new SampleConsensusModelLine<PointXYZ> (cloud.makeShared ()));
+
+  // Algorithm tests
+  pcl::Indices samples;
+  int iterations = 0;
+  model->getSamples(iterations, samples);
```

Comment:
I think having the same error message 1000 times in the test log is not so nice. How about changing this test by adding a third point unequal to the other two, and checking that this third point is in the sample?
Otherwise the changes look good :smile:

Reply:
Wouldn't that open up the possibility that the third point is selected on the first run by pure chance so the branch that this is supposed to test doesn't get exercised? That part of my brain got a little bit rusty, but if I'm not mistaken that could actually happen in 2 out of 3 cases (ignoring sample order).

Would it instead be an option to decrease the max. iteration count to something like 10 (technically one iteration would be sufficient, I guess :laughing:)?

Reply:
That would depend on the random number generation, especially the seed. By default, the behaviour is deterministic and the seed is always the same. So we should be able to find a configuration where that branch is always executed. Maybe the order of the points has to be switched (so that the "unequal" point is either at the beginning, or in the middle, or at the end). If we see that error message at least once in the test log, then we have a good configuration.
I also considered changing the max. number of iterations (`max_sample_checks_`), but that variable is `static const` with no way to change it from outside, and I would prefer to leave it at 1000.

Reply:
I likely had `setMaxIterations` of `SampleConsensus` in my mind, but of course we are in `SampleConsensusModel` here.

Parts of the test code below likely scream "DON'T!" right in your face, but I think that's about it what we can do here. Asserting on stdout/stderr does seem to be a [touchy topic](https://stackoverflow.com/a/33186201/) in googletest :laughing: Unfortunately It also doesn't work to assert on `iterations` being at least 2, since it is only set when there are not enough samples to draw from.

```cpp
TEST (SampleConsensusModelLine, SampleValidationPointsEqual)
{
  PointCloud<PointXYZ> cloud;
  cloud.resize (3);

  // The "cheat point" makes it possible to find a set of valid samples and
  // therefore avoids the log messages of an unsuccessful sample validation
  // being printed a 1000 times without any chance of success.
  // The order is chosen such that with a known, fixed rng-state/-seed all
  // validation steps are actually exercised.
  const pcl::index_t firstKnownEqualPoint = 0;
  const pcl::index_t secondKnownEqualPoint = 1;
  const pcl::index_t cheatPointIndex = 2;

  cloud[firstKnownEqualPoint].getVector3fMap () <<  0.1f,  0.0f,  0.0f;
  cloud[secondKnownEqualPoint].getVector3fMap () <<  0.1f,  0.0f,  0.0f;
  cloud[cheatPointIndex].getVector3fMap () <<  0.0f,  0.1f,  0.0f; // <-- cheat point

  // Create a shared line model pointer directly
  SampleConsensusModelLinePtr model (new SampleConsensusModelLine<PointXYZ> (cloud.makeShared ()));

  // Algorithm tests
  pcl::Indices samples;
  int iterations = 0;
  model->getSamples(iterations, samples);
  // The best option would be to assert on stderr output here, but that doesn't
  // seem to be that simple.
  EXPECT_EQ (samples.size(), 2);
  // The "cheat point" has to be part of the sample, otherwise something is wrong
  EXPECT_TRUE (std::find(samples.begin (), samples.end (), cheatPointIndex) != samples.end ());

  pcl::Indices forcedSamples = {firstKnownEqualPoint, secondKnownEqualPoint};
  Eigen::VectorXf modelCoefficients;
  EXPECT_FALSE (model->computeModelCoefficients (forcedSamples, modelCoefficients));
}
```

Reply:
Looks good. Maybe another comment that `model` is on purpose called with the default `random = false`

Reply:
I could also imagine going a step further and being even more explicit:
```cpp
// Create a shared line model pointer directly and explicitly disable the
// random seed for the reasons mentioned above
SampleConsensusModelLinePtr model (
  new SampleConsensusModelLine<PointXYZ> (cloud.makeShared (), /* random = */ false));
```
Just from some grep-ing, that "fake named argument" thing doesn't seem to be used widely in PCL.